rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,633 @@
+To configure your SSH client, add the following to your local SSH configuration
+(~/.ssh/config):
+
+Include /home/alice/tbot/ssh_config
+```
+
+Where possible, the informative text is written to stderr to allow easy
+appending:
+
+```
+$ tbot config ssh -c tbot.yaml >> ~/.ssh/config
+The following SSH config snippet was written to the pipe:
+
+Include /home/alice/tbot/ssh_config
+```
+
+Notes:
+ * This is a contrived example as the `ssh-client` template is always generated
+   by default for `ssh`-typed certificates, which are themselves the default
+   In other words, `ssh_config` and `known_hosts` are generated if no
+   `tbot.yaml` is specified.
+ * `tctl auth sign` has a tiny amount of configuration templating already for
+   Postgres and MongoDB certificates. We should aim to move this logic out into
+   a separate package that can render configuration snippets for a variety of
+   external systems. This work is out of scope for the purposes of this RFD.
+
+#### Polling for expiration
+
+To watch a certificate for expiration, the `teleport/lib/utils/interval`
+package will be used. The tbot will only set up polling for _user certs_, as
+host certificates issued by teleport do not expire (host certificates *will* be
+re-issued during CA rotations).
+
+#### Initial User Certificates
+
+In order for the bot to obtain its first set of user certificates, a new
+gRPC endpoint will be added to the API server which will mimic the host
+registration flow, accepting a pre-generated token and a public key to sign,
+but will use user tokens instead of provisioning tokens.
+
+Today, user tokens can be used to initiate the user invite flow, handle password
+resets, and manage the account recovery process. In order to generate initial
+user certificates, we will introduce a new type of user token for certificate
+renewal bots.
+
+When the backend receives the request to generate new user credentials, it validates
+the token, ensuring that the token is of the correct type and has not expired.
+
+This flow can be extended to support other auth methods beyond simple token
+registration; for example, we could additionally accept an EC2 identity
+document as is currently done for
+[Simplified Node Joining for AWS](0041-aws-node-join.md).
+
+#### Renewable User Certificates
+
+Teleport's current behavior when re-issuing a user certificate is maintain the
+original expiration time and refuse to bump out the TTL. We plan to preserve
+this behavior by default, and only allow for extending the expiration date on
+certificates that have been marked renewable.
+
+In order to mark a certificate as renewable, we'll include a new extension in
+the certificate's subject field. This attribute will only be set by teleport
+when the certificates are first issued using the new endpoint and user token.
+
+#### Audit Log
+
+The auth server will emit new events to the audit log when:
+
+- a new renewable certificate is issued for the first time
+- a certificate is renewed
```

Comment:
This event should contain:
* The final `destination` path that was set at the time
* A flag indicating the ACL set on the destination (if insecure)
* A flag indicating if the destination was a symlink, and what it resolved to
* Which certs were generated (`key`/`key.pub`/`sshcert`/`known_hosts`/`ssh_config`...) and eventually their pins/hashes

Reply:
As useful as information this info would be, there is currently no mechanism for relaying that sort of information to `generateUserCerts()`. The auth server also generates both cert types (TLS and SSH) for all cert requests, bot or otherwise.

I suppose we could extend the existing `UserCertsRequest` to add these as (optional) fields? It feels a bit odd to add fields expressly for audit logging purposes, but it is an option.